### PR TITLE
Serializes CensusContextImpl as a Protocol Buffer.

### PR DESCRIPTION
**Pull request by sebright**
This PR adds census-proto as a git submodule and then uses it for the `CensusContext` serialization format.

### CODE REVIEW

File: core/java/com/google/census/Census.java [hunk 1]
```diff
@@ -41,7 +41,11 @@ public static CensusContext getCurrent() {
     return CONTEXT_FACTORY.getCurrent();
   }
 
-  /** Creates a {@link CensusContext} from the given on-the-wire encoded representation. */
+  /**
+   * Creates a {@link CensusContext} from the given on-the-wire encoded representation. Should be
+   * the inverse of {@link CensusContext#serialize()}. The serialized representation should be
```
Comment (adriancole) [inline core/java/com/google/census/Census.java]: supernit: make a line break with <p> to separate summary description from details

File: core/java/com/google/census/CensusContext.java [hunk 2]
```diff
@@ -41,6 +41,8 @@
 
   /**
    * Serializes the {@link CensusContext} into the on-the-wire representation.
+   * The inverse of {@link CensusContextFactory#deserialize()} and should be based on the
```
Comment (adriancole) [inline core/java/com/google/census/CensusContext.java]: ditto about supernit and line break

File: core/java/com/google/census/CensusContextFactory.java [hunk 3]
```diff
@@ -14,14 +14,16 @@
...
-
 import javax.annotation.Nullable;
 
 /**
  * Factory class for {@link CensusContext}.
  */
 public interface CensusContextFactory {
-  /** Creates a new {@link CensusContext} built from the given on-the-wire encoded representation.
+  /**
+   * Creates a {@link CensusContext} from the given on-the-wire encoded representation. Should be
```
Comment (adriancole) [inline core/java/com/google/census/CensusContextFactory.java]: ditto

File: core_native/java/com/google/census/CensusSerializer.java [hunk 4]
```diff
@@ -36,27 +38,58 @@ static ByteBuffer serialize(CensusContextImpl context) {
...
-        int valIndex = input.indexOf(TAG_DELIM, keyIndex + 1);
-        String key = input.substring(keyIndex + 1, valIndex);
-        keyIndex = input.indexOf(TAG_PREFIX, valIndex + 1);
-        String val = input.substring(valIndex + 1, keyIndex == -1 ? input.length() : keyIndex);
-        tags.put(key, val);
-      } while (keyIndex != -1);
+  }
+
+  private static HashMap<String, String> fromString(String encoded) {
+    HashMap<String, String> tags = new HashMap<>();
```
Comment (adriancole) [inline core_native/java/com/google/census/CensusSerializer.java]: linkedhashmap would retain order. usually handy in tests

File: core_native/java/com/google/census/CensusSerializer.java [hunk 5]
```diff
@@ -36,27 +38,58 @@ static ByteBuffer serialize(CensusContextImpl context) {
...
-      do {
-        int valIndex = input.indexOf(TAG_DELIM, keyIndex + 1);
-        String key = input.substring(keyIndex + 1, valIndex);
-        keyIndex = input.indexOf(TAG_PREFIX, valIndex + 1);
-        String val = input.substring(valIndex + 1, keyIndex == -1 ? input.length() : keyIndex);
-        tags.put(key, val);
-      } while (keyIndex != -1);
+  }
+
+  private static HashMap<String, String> fromString(String encoded) {
```
Comment (adriancole) [inline core_native/java/com/google/census/CensusSerializer.java]: maybe rename to tagsFromString?

Comment (dinooliva) [inline core_native/java/com/google/census/CensusSerializer.java]: Makes sense.

File: core_native/java/com/google/census/CensusSerializer.java (hunk 4 above)
Comment (dinooliva) [inline core_native/java/com/google/census/CensusSerializer.java]: Agreed, that can be handy but I don't want to make it a requirement.

**Comment by sebright** [comment]
I rebased, formatted the Javadocs, and renamed CensusSerializer.fromString().

File: core/java/com/google/census/Census.java [hunk 6]
```diff
@@ -41,7 +41,12 @@ public static CensusContext getCurrent() {
     return CONTEXT_FACTORY.getCurrent();
   }
 
-  /** Creates a {@link CensusContext} from the given on-the-wire encoded representation. */
+  /**
+   * Creates a {@link CensusContext} from the given on-the-wire encoded representation.
+   *
+   * <p>Should be the inverse of {@link CensusContext#serialize()}. The serialized representation
```
Comment (dinooliva) [inline core/java/com/google/census/Census.java]: Consider putting the paragraph marker ("<p>") on the empty line above.

Comment (sebright) [inline core/java/com/google/census/Census.java]: The Google style guide specifies that `<p>` should be on the same line as the new paragraph: https://google.github.io/styleguide/javaguide.html#s7.1.2-javadoc-paragraphs

**Comment by sebright** [comment]
@adriancole @dinooliva  Thanks for the review!

**Pull request merged**